docs(api): refactor and rewrite Complex Commands (#13384)
OpenTrons/opentrons-api: api/docs/v2/complex_commands/parameters.rst

:og:description: Parameters for fine-tuning complex liquid handling behavior in the Python API.

.. _complex_params:

**********************************
Complex Liquid Handling Parameters
**********************************

Complex commands accept a number of optional parameters that give you greater control over the exact steps they perform.  

This page describes the accepted values and behavior of each parameter. The parameters are organized in the order that they first add a step. Some parameters, such as ``touch_tip``, add multiple steps. See :ref:`complex-command-order` for more details on the sequence of steps performed by complex commands.

> The API reference entry for :py:meth:`~.InstrumentContext.transfer` also lists the parameters and has more information on their implementation as keyword arguments.

Reviewer: Suggest removing tilde for the 1st reference to `InstrumentContext.transfer`.

.. _param-tip-handling:

Tip Handling
============

The ``new_tip`` parameter controls if and when complex commands pick up new tips from the pipette's tip racks. It has three possible values:

.. list-table::
   :header-rows: 1

   * - Value
     - Behavior
   * - ``"once"``

>      - This is the default behavior for all complex commands.

Reviewer: I'd recommend making this sentence a bullet too and even shorten it:

- Default behavior for all complex commands.
- Pick up a tip at the start of the command.
- Use the tip for all liquid handling.
- Drop the tip at the end of the command.

Making the 1st sentence a bullet makes the list uniform.

Author: 👍 uniformity within cell
😕 one of four bullets not describing actual robot movement

so i moved this after the table: 
```
`"once"`` is the default behavior for all complex commands.
```

   * - ``"never"``
     - Do not pick up or drop tips at all.

.. versionadded:: 2.0
     
Tip Handling Requirements
-------------------------
     
``"once"`` and ``"always"`` require that the pipette has an :ref:`associated tip rack <pipette-tip-racks>`, or the API will raise an error (because it doesn't know where to pick up a tip from). If the pipette already has a tip attached, the API will also raise an error when it tries to pick up a tip. 

.. code-block:: python

    pipette.pick_up_tip()
    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"], plate["B3"]],
        new_tip="never",  # "once", "always", or None will error
    )

Conversely, ``"never"`` requires that the pipette has picked up a tip, or the API will raise an error (because it will attempt to aspirate without a tip attached).

Avoiding Cross-Contamination
----------------------------

One reason to set ``new_tip="always"`` is to avoid cross-contamination between wells. However, you should always do a dry run of your protocol to test that the pipette is picking up and dropping tips in the way that your application requires.

> :py:meth:`~.InstrumentContext.transfer` will pick up a new tip before *every* aspirate when ``new_tip="always"``. This includes when :ref:`complex-tip-refilling` requires multiple aspirations from a single source well.

Reviewer: Same as another comment regarding using a page or section title in a sentence or going sentence case to make text fit visually into the sentence. Again, this is just aesthetics. 

```suggestion
:py:meth:`~.InstrumentContext.transfer` will pick up a new tip before *every* aspirate when ``new_tip="always"``. This includes when :ref:`refilling a tip <complex-tip-refilling>` requires multiple aspirations from a single source well.
```

    pipette.distribute(
        volume=200,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        new_tip="always",
    )
    
But it *does not* pick up a new tip after dispensing into B1:

.. code-block:: text

    Picking up tip from A1 of tip rack on 3
    Aspirating 220.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Dispensing 200.0 uL into B1 of well plate on 2 at 92.86 uL/sec
    Blowing out at A1 of Opentrons Fixed Trash on 12
    Aspirating 220.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Dispensing 200.0 uL into B2 of well plate on 2 at 92.86 uL/sec
    Blowing out at A1 of Opentrons Fixed Trash on 12
    Dropping tip into A1 of Opentrons Fixed Trash on 12

If this poses a contamination risk, you can work around it in a few ways:

    * Use ``transfer()`` with ``new_tip="always"`` instead.
    * Set :py:obj:`.well_bottom_clearance` high enough that the tip doesn't contact liquid in the destination well.

>     * Use :ref:`v2-atomic-commands` instead of complex commands.

Reviewer: ```suggestion
    * Use :ref:`building block commands <v2-atomic-commands>` instead of complex commands.
```

.. _param-mix-before:

Mix Before
==========

The ``mix_before`` parameter controls mixing in source wells before each aspiration. Its value must be a :py:class:`tuple` with two numeric values. The first value is the number of repetitions, and the second value is the amount of liquid to mix in µL.

For example, this transfer command will mix 50 µL of liquid 3 times before each of its aspirations::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        mix_before=(3, 50),
    )
    
.. versionadded:: 2.0

> Mixing occurs before every aspiration, including when :ref:`complex-tip-refilling` is required.

Reviewer: I'd recommend changing the link from the destination's title to sentence case.

```suggestion
Mixing occurs before every aspiration, including when :ref:`tip refilling <complex-tip-refilling>` is required.
```

.. note::
    :py:meth:`~.InstrumentContext.consolidate` ignores any value of ``mix_before``. Mixing on the second and subsequent aspirations of a consolidate command would defeat its purpose: to aspirate multiple times in a row, from different wells, *before* dispensing.
    
.. _param-disposal-volume:

Disposal Volume
===============

> The ``disposal_volume`` parameter controls how much extra liquid is aspirated as part of a :py:meth:`~.InstrumentContext.distribute` command. Including a disposal volume can improve the accuracy of each dispense. The pipette blows out the disposal volume of liquid after dispensing. To skip aspirating and blowing out extra liquid, set ``disposal_volume=0``.

Reviewer: This strikes me as an assertion made without any supporting explanation and I wonder if we can remove or improve it:

> Including a disposal volume can improve the accuracy of each dispense.

Might be more powerful to explain why:  "Including a disposal volume helps improve the accuracy of each dispense _because insert reasons here_. "

Author: support has this but i don't know whether it helps:
> In a multi-dispense sometimes a bit of liquid is lost. The Disposal volume is a buffer of extra liquid to make sure every destination well gets the full Transfer Volume.

…lost where? 😬

Reviewer: To the mystery of the ages perhaps? I see, that second sentence in the explanation helps. Not sure what to do here. Maybe just leave it be.

By default, ``disposal_volume`` is the :ref:`minimum volume <new-pipette-models>` of the pipette, but you can set it to any amount::

    pipette.distribute(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        disposal_volume=10,  # reduce from default 20 µL to 10 µL
    )
    
.. versionadded:: 2.0
    

> If the amount to aspirate plus the disposal volume exceeds the tip's capacity, ``distribute()`` will use a :ref:`complex-tip-refilling` strategy. In such cases, the pipette will aspirate and blow out the disposal volume *for each aspiration*. For example, this command will require tip refilling with a 1000 µL pipette::

Reviewer: Consider "tip refilling strategy" as the link instead of the page/section title case title.

        volume=120,
        source=reservoir["A1"],
        dest=[plate.columns()[0]],
        disposal_volume=50,
    )
    
The amount to dispense in the destination is 960 µL (120 µL for each of 8 wells in the column). Adding the 50 µL disposal volume exceeds the 1000 µL capacity of the tip. The command will be split across two aspirations, each with the full disposal volume of 50 µL. The pipette will dispose *a total of 100 µL* during the command.

.. note::
    :py:meth:`~.InstrumentContext.transfer` will not aspirate additional liquid if you set ``disposal_volume``. However, it will perform a very small blow out after each dispense.
    
    :py:meth:`~.InstrumentContext.consolidate` ignores ``disposal_volume`` completely.

.. _param-touch-tip:

Touch Tip
=========

The ``touch_tip`` parameter accepts a Boolean value. When ``True``, a touch tip step occurs after every aspirate and dispense.

For example, this transfer command aspirates, touches the tip at the source, dispenses, and touches the tip at the destination::

    pipette.transfer(
        volume=100,
        dest=plate["A1"],
        source=plate["B1"],
        touch_tip=True,
    )

.. versionadded:: 2.0

> Touch tip occurs after every aspiration, including when :ref:`complex-tip-refilling` is required.

Reviewer: ```suggestion
Touch tip occurs after every aspiration, including when :ref:`tip refilling <complex-tip-refilling>` is required.
```

> This parameter always uses default motion behavior for touch tip. Use the :ref:`touch tip building block command <touch-tip>` if you need to:

Reviewer: See, this is a great example of making the a title part of the sentence. Very nice.

    * Only touch the tip after aspirating or dispensing, but not both.
    * Control the speed, radius, or height of the touch tip motion.

.. _param-air-gap:

Air Gap
=======

The ``air_gap`` parameter controls how much air to aspirate and hold in the bottom of the tip when it contains liquid. The parameter's value is the amount of air to aspirate in µL.

Air-gapping behavior is different for each complex command. The different behaviors all serve the same purpose, which is to never leave the pipette holding liquid at the very bottom of the tip. This helps keep liquids from seeping out of the pipette.

> .. list-table::

Reviewer: Such an _awesome_ table. I love it.

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=plate["B1"],
        air_gap=20,
    )
    
.. versionadded:: 2.0

When consolidating, air gaps still occur after every aspiration. In this example, the tip will use 210 µL of its capacity (50 µL of liquid followed by 20 µL of air, repeated three times)::

    pipette.consolidate(
        volume=50,
        source=[plate["A1"], plate["A2"], plate["A3"]],
        dest=plate["B1"],
        air_gap=20,
    )

.. code-block:: text

    Picking up tip from A1 of tip rack on 3
    Aspirating 50.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Air gap
        Aspirating 20.0 uL from A1 of well plate on 2 at 92.86 uL/sec
    Aspirating 50.0 uL from A2 of well plate on 2 at 92.86 uL/sec
    Air gap
        Aspirating 20.0 uL from A2 of well plate on 2 at 92.86 uL/sec
    Aspirating 50.0 uL from A3 of well plate on 2 at 92.86 uL/sec
    Air gap
        Aspirating 20.0 uL from A3 of well plate on 2 at 92.86 uL/sec
    Dispensing 210.0 uL into B1 of well plate on 2 at 92.86 uL/sec
    Dropping tip into A1 of Opentrons Fixed Trash on 12
    

> If adding an air gap would exceed the pipette's maximum volume, the complex command will use a :ref:`complex-tip-refilling` strategy. For example, this command uses a 300 µL pipette to transfer 300 µL of liquid plus an air gap::

Reviewer: ```suggestion
If adding an air gap would exceed the pipette's maximum volume, the complex command will use a :ref:`tip refilling <complex-tip-refilling>` strategy. For example, this command uses a 300 µL pipette to transfer 300 µL of liquid plus an air gap::
```

	Dispensing 170.0 uL into B1 of well plate on 2 at 92.86 uL/sec
	Aspirating 150.0 uL from A1 of well plate on 2 at 92.86 uL/sec
	Air gap
		Aspirating 20.0 uL from A1 of well plate on 2 at 92.86 uL/sec
	Dispensing 170.0 uL into B1 of well plate on 2 at 92.86 uL/sec
	Dropping tip into A1 of Opentrons Fixed Trash on 12

.. _param-mix-after:

Mix After
=========

The ``mix_after`` parameter controls mixing in source wells after each dispense. Its value must be a :py:class:`tuple` with two numeric values. The first value is the number of repetitions, and the second value is the amount of liquid to mix in µL.

For example, this transfer command will mix 50 µL of liquid 3 times after each of its dispenses::

    pipette.transfer(
        volume=100,
        source=plate["A1"],
        dest=[plate["B1"], plate["B2"]],
        mix_after=(3, 50),
    )
    
.. versionadded:: 2.0

.. note::
    :py:meth:`~.InstrumentContext.distribute` ignores any value of ``mix_after``. Mixing after dispensing would combine (and potentially contaminate) the remaining source liquid with liquid present at the destination.

.. _param-blow-out:

Blow Out
========

> There are two parameters that control blowout behavior. The ``blow_out`` parameter accepts a Boolean value. When ``True``, the pipette blows out remaining liquid when the tip is empty or only contains the disposal volume. The ``blowout_location`` parameter controls in which of three locations these blowout actions occur. The default blowout location is the trash. Blowout behavior is different for each complex command. 

Reviewer: Ed, I had a struggle with this paragraph. It says `blow_out` accepts 2 parameters that control behavior, but one is a Boolean, so that is already `true | false`, so now we have 3. I was confused about what the parameters actually are. How about trying bullets to make the 2 parameters more clear. Something like this maybe:

```suggestion
There are two parameters that control blowout behavior. One is a Boolean ``True`` and the other is ``blowout_location``.

- When ``blow_out=True``, the pipette expels any remaining liquid when the tip is empty or contains the disposal volume.
- The ``blowout_location`` parameter tells the robot where to perform a blowout action. The default blowout location is the trash but there are different location options for each complex command.
```

And then lead into/keep the table that follows.

Author: The parameters are the names, not their possible values: `blow_out` and `blowout_location`.

Perhaps a clearer intro sentence solves the issue without restructuring the paragraph, like `There are two parameters that control whether and where the pipette blows out liquid.`

Reviewer: Yes!! Much better.

.. list-table::
   :header-rows: 1

   * - Method

>      - Blowout behavior

Reviewer: I'm always unsure about how we'd like to capitalize multiple words in table column headers: title or sentence case.

Also, as the 2nd column deals with 2 items, I'd recommend changing the header title to "Blowout Behavior and Location" from "Blowout behavior". I think there's enough space that the extra text won't make the table wider.

Author: I was wrapping all of those concepts into the word "behavior", but I think it does make sense to have two nouns in the header since the contents of the column describe two parameters.

   * - ``transfer()``
     -
       - Blow out after each dispense.
       - Valid locations: ``"trash"``, ``"source well"``, ``"destination well"``
   * - ``distribute()``
     - 
       - Blow out after the final dispense.
       - Valid locations: ``"trash"``, ``"source well"``
   * - ``consolidate()``
     - 
       - Blow out after the only dispense.
       - Valid locations: ``"trash"``, ``"destination well"``

For example, this transfer command will blow out liquid in the trash twice, once after each dispense into a destination well::

    pipette.transfer(
        volume=100,
        source=[plate["A1"], plate["A2"]],
        dest=[plate["B1"], plate["B2"]],
        blow_out=True,
    )

> 

Reviewer: Could we add a second example? The first one shows the default location behavior. A second one could show a specific location. My guess at this is:

```
pipette.transfer(
        volume=100
        source[plate["A1"], plate["A2"]],
        dest=[plate["B1"], plate["B2"]],
        blow_out=True,
        blowout_location="destination well",
)
```

And then mirror the intro sentence text, but add that the blowout location is in wells B1 and B2.

Author: These examples come after the note. Do you think it improves the flow if we move the note down?

Reviewer: Yes, moving the note helps. Now I see the default code behavior and the `blowout_location` examples. More obvious. Thanks!

.. versionadded:: 2.0

.. note::

>     If the tip already contains liquid before the complex command, the default blowout location will shift away from the trash. Transfer and distribute shift to the source well, and consolidate shifts to the destination well. For example, this transfer command will blow out in well B1 because it's the source::

Reviewer: ```suggestion
    If the tip already contains liquid before the complex command, the default blowout location will shift away from the trash. ``transfer`` and ``distribute`` shift to the source well. ``consolidate`` shifts to the destination well. For example, this transfer command will blow out in well B1 because it's the source::
```

Author: will do, plus `()` on each
